core/src/main/java/io/questdb/griffin/engine/functions/finance/LevelTwoPriceFunctionFactory.java

```diff
@@ -0,0 +1,336 @@
+
+            final double size_1 = args.getQuick(3).getDouble(rec);
+            final double value_1 = args.getQuick(4).getDouble(rec);
+
+            if (size_0 + size_1 >= target) {
+                return ((size_0 * value_0) + (target - size_0) * value_1) / target;
+            }
+
+            final double size_2 = args.getQuick(5).getDouble(rec);
+            final double value_2 = args.getQuick(6).getDouble(rec);
+
+            if (size_0 + size_1 + size_2 >= target) {
+                return ((size_0 * value_0)
+                        + (size_1 * value_1)
+                        + (target - size_0 - size_1) * value_2) / target;
+            }
+
+            final double size_3 = args.getQuick(7).getDouble(rec);
+            final double value_3 = args.getQuick(8).getDouble(rec);
+
+            if (size_0 + size_1 + size_2 + size_3 >= target) {
+                return ((size_0 * value_0)
+                        + (size_1 * value_1)
+                        + (size_2 * value_2)
+                        + (target - size_0 - size_1 - size_2) * value_3) / target;
+            }
+
+            final double size_4 = args.getQuick(9).getDouble(rec);
+            final double value_4 = args.getQuick(10).getDouble(rec);
+
+            if (size_0 + size_1 + size_2 + size_3 + size_4 >= target) {
+                return ((size_0 * value_0)
+                        + (size_1 * value_1)
+                        + (size_2 * value_2)
+                        + (size_3 * value_3)
+                        + (target - size_0 - size_1 - size_2 - size_3) * value_4) / target;
+            }
+
+            return Double.NaN;
+        }
+    }
+
+    private static class L2PriceFunctionN extends L2PriceBaseFunction {
+        public L2PriceFunctionN(ObjList<Function> args) {
+            super(args);
+        }
+
+        @Override
+        public double getDouble(Record rec) {
+            final long t = args.getQuick(0).getLong(rec);
+
+            double ta = 0; // target accumulator
+            double pa = 0; // price accumulator
+            double rt = t; // reduced target
+            double pp; // partial price
+
+            // expect (size, value) pairs
+            // equation is
+            // ((size[0] * value[0]) + (size[n] * value[n]) + ((target - size[0] - size[n]) * value[n+1])) / target
+            // ((ra)             + (rt * value[n+1])) / t
+            // get final price by partially filling against the last bin
+            for (int i = 1, n = args.size(); i < n; i += 2) {
+                final double size = args.getQuick(i).getDouble(rec);
+                final double value = args.getQuick(i + 1).getDouble(rec);
+
+                // add size to acc
+                ta += size;
+
+                // if order not fulfilled
+                if (ta < t) {
+                    pa += (size * value);
+                    rt -= size;
+                    continue;
+                }
+
+                if (ta >= t) {
```

Comment:
This `if` is not necessary as far as I can see. I'd rewrite the `if`s like this instead

```java
if (ta >= t) {
                   if (i == 1) {
                        return value;
                    }

                    pp = rt * value;
                    return (pa + pp) / t;
} else {
                    pa += (size * value);
                    rt -= size;
}